syntax = "proto2";

import "proto/animation.proto";
import "proto/event.proto";
import "proto/primitives.proto";
import "proto/scene-node.proto";

package troll;

message Action {
  oneof Action {
    NoopAction noop = 1;
    QuitAction quit = 2;
    EmitAction emit = 3;

    ChangeSceneAction change_scene = 4;

    SceneNodeAction create_scene_node = 5;
    SceneNodeAction destroy_scene_node = 6;

    CollisionAction on_collision = 7;
    CollisionAction on_detaching = 8;

    AnimationScriptAction play_animation_script = 9;
    AnimationScriptAction stop_animation_script = 10;
    AnimationScriptAction pause_animation_script = 11;

    DisplayTextAction display_text = 12;

    ScriptAction import_module = 13;
    ScriptAction call = 14;
  }
}

message NoopAction {}

message QuitAction {}

message EmitAction {
  optional Event event = 1;
}

message ChangeSceneAction {
  optional string scene_id = 1;
}

message SceneNodeAction {
  optional SceneNode scene_node = 1;
}

message CollisionAction {
  repeated string sprite_id = 1;
  repeated string scene_node_id = 2;
  repeated Action action = 3;
}

message AnimationScriptAction {
  // Fields script and script_id are mutually exclusive. If a script exists the
  // script_id is ignored for playing a new animation. For stop/pause animations
  // only the script_id field is relevant.
  optional AnimationScript script = 1;
  optional string script_id = 2;
  
  optional string scene_node_id = 3;
}

message DisplayTextAction {
  optional string text = 1;
  optional Vector position = 2;
  optional int32 display_duration = 3;

  optional int32 size = 4;
  optional bool bold = 5;
  optional bool italics = 6;

  optional RGBa colour = 7;
  optional RGBa background_colour = 8;
  optional string font = 9;
}

message ScriptAction {
  optional string module = 1;
  optional string function = 2;
}
